core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java

```diff
@@ -0,0 +1,190 @@
+            label, config.timeoutCheckIntervalMillis(), groupManager, groupCleanupWorker);
+    this.groupCloseWorker =
+        new GroupCloseWorker<>(
+            label, config.timeoutCheckIntervalMillis(), delayedSlotMoveWorker, groupCleanupWorker);
+    this.groupManager.setGroupCloseWorker(groupCloseWorker);
+    this.groupManager.setGroupCleanupWorker(groupCleanupWorker);
+
+    // TODO: This should be replaced by other metrics mechanism.
+    this.monitorExecutorService =
+        Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat(label + "-group-commit-monitor-%d")
+                .build());
+    startMonitorExecutorService();
+  }
+
+  Metrics getMetrics() {
+    return new Metrics(
+        groupCloseWorker.size(),
+        delayedSlotMoveWorker.size(),
+        groupCleanupWorker.size(),
+        groupManager.sizeOfNormalGroupMap(),
+        groupManager.sizeOfDelayedGroupMap());
+  }
+
+  // TODO: This should be replaced by other metrics mechanism.
+  private void startMonitorExecutorService() {
+    Runnable print =
+        () -> logger.info("[MONITOR] Timestamp={}, Metrics={}", Instant.now(), getMetrics());
+
+    monitorExecutorService.execute(
+        () -> {
+          while (!monitorExecutorService.isShutdown()) {
+            print.run();
+            Uninterruptibles.sleepUninterruptibly(1, TimeUnit.SECONDS);
+          }
+          print.run();
+        });
+  }
+
+  /**
+   * Set an emitter which contains implementation to emit values. Ideally, this should be passed to
+   * the constructor, but the instantiation timings of {@link GroupCommitter} and {@link Emittable}
+   * can be different. That's why this API exists.
+   *
+   * @param emitter An emitter.
+   */
+  public void setEmitter(Emittable<EMIT_KEY, V> emitter) {
+    groupManager.setEmitter(emitter);
+  }
+
+  /**
+   * Reserves a new slot in the current {@link NormalGroup}. The slot may be moved to a {@link
+   * DelayedGroup} later.
+   *
+   * @param childKey A child key.
+   * @return The full key associated with the reserved slot.
+   */
+  public FULL_KEY reserve(CHILD_KEY childKey) {
+    while (true) {
+      FULL_KEY fullKey = groupManager.reserveNewSlot(childKey);
+      if (fullKey != null) {
+        return fullKey;
+      }
+      logger.debug(
+          "Failed to reserve a new value slot since the group was already closed. Retrying. key:{}",
+          childKey);
+    }
+  }
+
+  /**
+   * Marks the slot associated with the specified key READY and then, waits until the group which
+   * contains the slot is emitted.
+   *
+   * @param fullKey A full key associated with the slot already reserved with {@link
+   *     GroupCommitter#reserve}.
+   * @param value A value to be set to the slot.
+   * @throws GroupCommitException when group commit fails
+   */
+  public void ready(FULL_KEY fullKey, V value) throws GroupCommitException {
```

Comment:
What is the purpose of this `value` argument? Maybe we use it for a transaction ID (a child key) for each transaction within a group?

Reply:
In the case of coordinator state commit, `V` is supposed to be `Snapshot` and a passed value will be committed with other snapshots in the same group.

Reply:
Updated the comment to make it clearer in https://github.com/scalar-labs/scalardb/pull/1607/commits/fa5e72631b3ca1ef5eee6a47fff13071d9adbd12